core/src/mender-client.c

```diff
@@ -805,6 +801,77 @@ mender_client_initialization_work_function(void) {
     return ret;
 }
 
+static mender_err_t
+mender_commit_artifact_data(void) {
+
+    /* Load `artifact_name` from deployment - if we use provides-depends, then we
+     * use `artifact_name` from the provides instead. Before storing provides, we
+     * need to remove the artifact_name and store it in a separate key
+     * */
+    char *artifact_name = NULL;
+
+    cJSON *json_artifact_name = cJSON_GetObjectItemCaseSensitive(mender_client_deployment_data, "artifact_name");
+    if (NULL == json_artifact_name) {
+        mender_log_error("Unable to get artifact name from the deployment data");
+        return MENDER_FAIL;
+    }
+
+    artifact_name = json_artifact_name->valuestring;
+
+#ifdef CONFIG_MENDER_FULL_PARSE_ARTIFACT
+#ifdef CONFIG_MENDER_PROVIDES_DEPENDS
+    /* Get provides from the deployment data */
+    cJSON *provides = cJSON_GetObjectItemCaseSensitive(mender_client_deployment_data, "provides");
+    if (NULL == provides) {
+        mender_log_error("Unable to get new_provides from the deployment data");
+        return MENDER_FAIL;
+    }
+
+    /* Parse provides */
+    mender_key_value_list_t *new_provides = NULL;
+    if (MENDER_OK != mender_utils_string_to_key_value_list(provides->valuestring, &new_provides)) {
+        mender_log_error("Unable to parse provides from the deployment data");
+        return MENDER_FAIL;
+    }
+
+    /* Artifact name from provides takes precedence */
+    for (mender_key_value_list_t *item = new_provides; NULL != item; item = item->next) {
+        if (StringEqual("artifact_name", item->key)) {
+            artifact_name = item->value;
+        }
+    }
+#endif /* CONFIG_MENDER_PROVIDES_DEPENDS */
+#endif /* CONFIG_MENDER_FULL_PARSE_ARTIFACT */
+
+    mender_key_value_list_t *new_artifact_name = NULL;
+    if (MENDER_OK != mender_utils_create_key_value_node("artifact_name", artifact_name, &new_artifact_name)) {
+        mender_log_error("Unable to create key-value node for artifact name");
+        return MENDER_FAIL;
+    }
+
+    if (MENDER_OK != mender_storage_set_artifact_name(new_artifact_name)) {
+        mender_log_error("Unable to set artifact name");
+        return MENDER_FAIL;
+    }
+
+#ifdef CONFIG_MENDER_FULL_PARSE_ARTIFACT
+#ifdef CONFIG_MENDER_PROVIDES_DEPENDS
+    /* Delete artifact_name from the provides list */
+    if (MENDER_OK != mender_utils_key_value_list_delete_node(&new_provides, "artifact_name")) {
+        return MENDER_FAIL;
+    }
+
+    /* Replace the stored provides with the new provides */
+    if (MENDER_OK != mender_storage_set_provides(new_provides)) {
+        mender_log_error("Unable to set provides");
+        return MENDER_FAIL;
+    }
+#endif /* CONFIG_MENDER_PROVIDES_DEPENDS */
+#endif /* CONFIG_MENDER_FULL_PARSE_ARTIFACT */
+
+    return MENDER_OK;
```

Comment:
What does destroy `provides` and `new_provides`?

Reply:
Nothing, overlooked it